proxy/proxy/client.go

```diff
@@ -328,6 +333,30 @@ func (c *Client) tryConnect(addr string, cfg *tls.Config) (net.Conn, error) {
 	return ret, nil
 }
 
+// dialer selects the dialer to use from the Client configuration
+func (c *Client) dialer() ContextDialer {
```

Comment:
I just extracted it to make `tryConnect` (opinionatedly) cleaner.

Moved the logic back there and removed `c.dialer()` at all. 👍

Reply:
FWIW I disagree with Kurtis here. It's his call ultimately, and since the code is currently like this feel free to leave it, but: this tryConnect function is quite complicated in part because there's no clear separation of concerns: it's currently in charge of a lot of things. I think separating the code out into a separate function which is in charge of obtaining the dialer is a better decision (after this change, not before) because this operation now has a lot of branching and decisions - even if it's just used once, it's good to organize complicated code so that there's a clear API that future maintainers can wrap their head around.

Also, I think the original name was good: it's quite common to have a field named X and a method named x which accesses that field and does something special with it (and/or other fields). The similar naming is an intentional nod to the fact that, yes, the `dialer` method probably accesses the `Dialer` field. I'd only find it confusing if the `dialer` method didn't access the `Dialer` field, I guess.  But any other name works too, I don't really care what it's called as long as it's not part of the exported API.

(Also, I can't see the original change because it was force-pushed, so I'm just guessing what it looked like; please prefer to not force-push, as it makes code reviews harder to do because it's harder to see the full history after the fact.)

Reply:
So the root of my concern here is that the public API for this package is a bit less well defined (or at least understood by myself) and I'm not entirely sure if use of the `Client` struct is reasonable accessed by any downstream dependencies outside of this package. 

With that in mind, it seems confusing to me to have a public (exported) `*Client.Dialer` and `*Client.DialerContext` attributes, but then have a private (non-exported) `c *Client.dialer() ContextDialer` determining which are used. I wanted to avoid this function being exported at some point in the future with the context of the conversation lost and leading to a confusing or possible changed API surface. 

So my objection is less about where the code is structurally organized, and more of trying to keep the overall namespace a little more clear.

Reply:
Hey, the original commit can still be seen in the PR timeline in github: https://github.com/GoogleCloudPlatform/cloudsql-proxy/commit/68542d0e4a6fc6b98a73a5741169cd3bd82d614e

Wasn't sure about force pushing, or adding a `refactor:` commit (refactoring something I'm proposing? Then I'm forcing to squash) but yeah, I see your point.

Reply:
We can always `git checkout 68542d0 - proxy/proxy/client.go && git commit proxy/proxy/client.go -m "refactor: Extract dialer selection logic into Client.dialer()` :)

Reply:
I've extract back the logic in https://github.com/GoogleCloudPlatform/cloudsql-proxy/pull/483/commits/393b14a9bf4fa195de47ca164a8844ddd852219c in order to keep `tryConnect` clean, and I renamed the function into `selectDialer()` to address @kurtisvg's concerns about colliding naming.

Reply:
> I wanted to avoid this function being exported at some point in the future with the context of the conversation lost and leading to a confusing or possible changed API surface.

Note that it would be a compile error to have a Dialer method on a struct with a Dialer field; so I think this is not actually a problem.

selectDialer seems like an OK name too.

Reply:
> Hey, the original commit can still be seen in the PR timeline in github: 68542d0

Ah thanks, I didn't think to look there. I clicked on the conversation and github claimed that the code it was on didn't exist, so I trusted the web UI :)

As for squashing, all the open source repositories I'm involved with prefer to squash-and-commit at the end, so that it doesn't matter how many commits it took during review. This keeps the main tree's history clean (one commit per PR), while still allowing individual commits to be seen in the PR itself if anyone is interested in the history.

Reply:
I prefer an intact history via separate commits, but greatly prefer use rebasing over merge when updating upstream - but that particular bikeshed doesn't need to be painted in this thread :)